### ISSUE TITLE
MPI support

### ISSUE THREAD

**Issue by fredRos** [pyFFTW/pyFFTW#38]
I checked out pyFFTW, and it works flawlessly using multiple threads. But for very demanding applications on a super computer, I would need support for the MPI version of fftw. Are there any plans to include this into pyFFTW? What would one need to do? I could rally up some support locally to implement this but some guidance on how to start would greatly appreciated.

**Comment by fredRos** [comment]
Time to revive this issue. I forked pyfftw last year and wrote an MPI wrapper. I couldn't really come with an elegant solution that reuses most of the code @hgomersal wrote. In particular, building was a little more demanding. I made sure to have unit tests covering the various options though I know it is not as thorough as the the suite Henry built. 

pyfftw + MPI code is used in the development version of nifty, https://github.com/information-field-theory/nifty/tree/develop. 

I am fully aware that at present the code is fully usable but there is still a lot of work to be done until it could be merged into the master branch. I would really love to see it there because I already invested a lot of time into this, and it seems more people want this. Right now, I support 1D and multidimensional FFT with r2c, c2r, and c2c.

 I focused on writing a wrapper that is safe and easy to use from python. All the stuff fftw expects you to remember when using it is hidden behind 2 function and one class. I do not have any automatic scattering of data implemented. Currently I think that it's not needed. If somebody really needs MPI because of an expensive calculation on multiple nodes, then that person has to have a way of scattering the data anyways. And then I don't see how to hide the way fftw does it work without sacrificing performance.

I will send a pull request to get this going. I myself wouldn't accept it at present but this is to pave the way for future inclusion. Henry can tell me about what documentation and tests he wants.

**Comment by hgomersall** [comment]
Thanks Frederick, this looks fantastic. Bear with me a few days. I'm just on holiday and have a few other things to get sorted too. Cheers, Henry

**Comment by hgomersall** [comment]
Sorry for letting this fall by the wayside. Is this patch still a work in progress?

**Comment by fredRos** [comment]
Yes, I didn't change anything since I started #67 . I suggest you have a quick look at where I implemented things and how I had to change `setup.py`. I'm sure you made lots of changes since I forked, so there are plenty of merge conflicts. We can then have a discussion by skype or google+ on how to get my changes ready to be merged.

**Comment by mikaem** [comment]
Hi

Just wanted to let you know that I've been using pyfftw and mpi4py to do FFTs in parallel with MPI (http://arxiv.org/abs/1602.03638), with thousands of processors on a Blue Gene supercomputer. Using pyfftw to do the serial FFT and then mpi4py to do the required MPI communications does not really require much work and if you are already working in Python, there's really no good reason why you should have to wrap the FFTWs MPI routines. In fact, four lines of compact Python code is all you need to execute an FFT with slab decomposition, which is exactly what FFTW does. I've implemented it here: https://github.com/spectralDNS/mpiFFT4py. Since I'm a researcher the code is tailored to my needs, which is a pseudo-spectral direct numerical simulation code for turbulence, but perhaps it can be of use to you as well?

Best regards

Mikael

**Comment by rainwoodman** [comment]
Hi Mikael, I really liked your approach!

**Comment by mikaem** [comment]
Thanks:-)

**Comment by drwells** [comment]
I originally wrote the real to real bindings for some pseudospectral code. I will give this a look!

I am not doing much with pseudospectral methods any more but I still like them.

**Comment by mikaem** [comment]
Hi @drwells 

I'm actually using your branch of pyfftw with the dct:-) Why has this not been moved into the main rep yet?

**Comment by drwells** [comment]
@mikaem A large number of reasons. The project I originally wrote those for never really went anywhere for a variety of reasons; primarily my PhD research was somewhere else and I wanted to graduate, and that required finishing other things.

Fortunately I have cleaned up the branch a lot and, depending on what @hgomersall wants, #95 will be merged in the near future :)

**Comment by mikaem** [comment]
Ah. Wanting to graduate.. Wish some of my students were that mature (and not reading my github feeds) Hope to see the PR merged soon:-)

**Comment by ultimanet** [comment]
Hi @hgomersall !

any news regarding merging #67 into master? @fredRos's mpi branch is already heavily used here at various Max-Planck institutes and admins start to install it as the default thing. But in my opinion it would be cleaner if they installed the master branch, of course. 

@mikaem Does mpiFFT4 support n-dimensional transforms? A quick look just revealed decompositions for up to 3 dimensions; but maybe I missed something. 

Cheers!

**Comment by mikaem** [comment]
Hi @ultimanet 
[mpiFFT4py](https://github.com/spectralDNS/mpiFFT4py) is only for transforms up to 3D. I don't think it would be very difficult to extend it, but that is not currently on my todo-list.

**Comment by rainwoodman** [comment]
My two comments on #67.

1) if one insists the fftw library is bundled with then pyFFTW package,
then a lot of detection logic in setting up the package can be avoided --
is current pyFFTW pip install friendly? #67 seems to be not.

2) The patch seems to be adding a communicator to the original FFTW class,
but not using it. Putting a lot of logic in a pxi file is also slightly
concerning. Does FFTW_MPI really require the internals of FFTW class? Does
this indicate some members of FFTW class shall be exposed via Python API
but not?

**Comment by fredRos** [comment]
Back from a long vacation...



> 1) if one insists the fftw library is bundled with then pyFFTW package, then a lot of detection logic in setting up the package can be avoided -- is current pyFFTW pip install friendly? #67 seems to be not.

It is not entirely pip friendly because I need `mpicxx` instead of the usual `g++` or `clang`. Also I have not tested it at all on windows because I have no experience with programming on that platform.

Regarding detection logic, I see this is independent of supporting mpi or not, as shown by the recurrence of issues such #27 and #70. I had to compile FFTW myself on a cluster due to compiler dependencies and had trouble creating a long precision variant, so I had to modify the pyfftw build to detect what is there. There are even more variants to support, for example in the debian repo I find a quad precision, on a supercomputer I have seen variants with openmp or posix threads etc. I admit I sort of hacked it together in my build but it may provide a basis to solve #27 and #70. The python setuptools support needed is very basic, so I had to implement quite a lot by myself.

I'd say roughly 25 % of the mpi interface is in conditional compilation clauses. Going back to require single, double, and long is easily done by remove all `IF HAVE_*` statements. But should #27 or #70 be tackled, it would be quite some pain to reintroduce them in the right places.


> The patch seems to be adding a communicator to the original FFTW class, but not using it. Putting a lot of logic in a pxi file is also slightly concerning. 

good catch. The comm is a relict I need to remove. I first modified the original FFTW class, then moved everything out into a separate pxi file to keep it clear what's new and what's not. I think this helps a lot with the desire for manageable chunks to merge as expressed above. Should it remain in the pxi in the future? I'd let @hgomersall decide.


> Does FFTW_MPI really require the internals of FFTW class? Does this indicate some members of FFTW class shall be exposed via Python API but not?

I tried to make `FFTW_MPI` as similar as possible to the `FFTW` class. But `FFTW_MPI` obviously has more features that `FFTW` but also lacks some (e.g. in the guru interface) because they simply don't exist in the MPI version of the C FFTW library so I found no other way but ugly duplication and modification of details. With this, there is no need to expose anything via an API.
